### PR TITLE
http: fixing upgrade response bug

### LINKED ISSUE #10566: Envoy produces invalid response when TLS http/1.1 cluster returns `upgrade: h2` and `transfer-encoding: chunked`
Envoy produces invalid response when TLS http/1.1 cluster returns `upgrade: h2` and `transfer-encoding: chunked`

When http/1.1 TLS cluster returns following headers in a response:
```
Connection: upgrade
Upgrade: h2
Transfer-Encoding: chunked
```
Envoy returns invalid response downstream.

Direct request to the upstream using curl works correctly:
```
#> curl -k --http1.1 -v https://localhost:8443/upgrade-h2-chunked
...
< HTTP/1.1 200 OK
< Connection: upgrade
< Upgrade: h2
< Date: Sun, 29 Mar 2020 17:18:13 GMT
< Transfer-Encoding: chunked
<
Service response /upgrade-h2-chunked
```

Through Envoy we receive something like this:
```
#> curl -H "host: service" -v http://localhost:6002/upgrade-h2-chunked
< HTTP/1.1 200 OK
< upgrade: h2
< date: Sun, 29 Mar 2020 17:18:31 GMT
< x-envoy-upstream-service-time: 0
< server: envoy
< transfer-encoding: chunked
<
25
Service response /upgrade-h2-chunked

0

* transfer closed with outstanding read data remaining
* Closing connection 0
curl: (18) transfer closed with outstanding read data remaining
```

Responses with `transfer-encoding: chunked` only or `upgrade: h2` only work correctly through Envoy.

It seems like Envoy doesn't decode chunked encoding correctly when `upgrade: h2` is present. The strange numbers added to an original response are simply not decoded chunked encoding metadata.

If I see correctly, Envoy doesn't support Upgrade to http2:
> Currently, Envoy only supports prior knowledge for upstream connections 

[docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#config-cluster-v3-cluster)

Even if Envoy doesn't support it, I think an `Upgrade` header in a response with status other than 426 should be treated as an advertisment and can be safely ignored:
> A server MAY send an Upgrade header field in any other response to
   advertise that it implements support for upgrading to the listed
   protocols, in order of descending preference, when appropriate for a
   future request. (rfc7230)

Somebody already reported similar problem on Istio github: https://github.com/istio/istio/issues/17790

In real life case we encountered this problem when upstream returned `upgrade: h2,h2c`, so let's not forget about that case too.

## How to reproduce

I prepared docker-compose environment where this error can be reproduced. More info in README: https://github.com/MarcinFalkowski/envoy-broken-response-reproduce
The problem can be reproduced on latest master version (commit a29664c3a8dfa2337be0781c1cee747aa149f482)

## Logs

In Envoy logs we can see that Envoy enters `upgrade mode` and then wait for upstream for additional data (despite fact, that all data are already sent) until timeout occurs
<details>
<summary>envoy logs</summary>

```
envoy_1    | [2020-03-29 16:51:21.858][13][trace][http] [source/common/http/http1/codec_impl.cc:636] [C1] message begin
envoy_1    | [2020-03-29 16:51:21.858][13][trace][http] [source/common/http/http1/codec_impl.cc:425] [C1] completed header: key=Connection value=upgrade
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:425] [C1] completed header: key=Upgrade value=h2
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:425] [C1] completed header: key=Date value=Sun, 29 Mar 2020 16:51:21 GMT
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:559] [C1] onHeadersCompleteBase
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:425] [C1] completed header: key=Transfer-Encoding value=chunked
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:589] [C1] codec entering upgrade mode.
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:924] [C1] Client: onHeadersComplete size=4
envoy_1    | [2020-03-29 16:51:21.859][13][debug][router] [source/common/router/router.cc:1142] [C0][S9047571861382008090] upstream headers complete: end_stream=false
envoy_1    | [2020-03-29 16:51:21.859][13][debug][http] [source/common/http/conn_manager_impl.cc:1702] [C0][S9047571861382008090] encoding headers via codec (end_stream=false):
envoy_1    | ':status', '200'
envoy_1    | 'upgrade', 'h2'
envoy_1    | 'date', 'Sun, 29 Mar 2020 16:51:21 GMT'
envoy_1    | 'x-envoy-upstream-service-time', '6'
envoy_1    | 'server', 'envoy'
envoy_1    |
envoy_1    | [2020-03-29 16:51:21.859][13][trace][connection] [source/common/network/connection_impl.cc:429] [C0] writing 146 bytes, end_stream false
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:615] [C1] message complete
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:621] [C1] Pausing parser due to upgrade.
envoy_1    | [2020-03-29 16:51:21.859][13][trace][http] [source/common/http/http1/codec_impl.cc:483] [C1] parsed 118 bytes

...

envoy_1    | [2020-03-29 16:51:36.851][13][debug][router] [source/common/router/router.cc:812] [C0][S9047571861382008090] upstream timeout
envoy_1    | [2020-03-29 16:51:36.851][13][debug][router] [source/common/router/upstream_request.cc:264] [C0][S9047571861382008090] resetting pool request
```

</details>

## Config

<details>
<summary>cluster</summary>

```yaml
  clusters:
  - name: service
    type: STRICT_DNS
    connect_timeout: 1s
    load_assignment:
      cluster_name: service
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: service
                port_value: 443
    transport_socket:
      name: envoy.transport_sockets.tls
```

</details>

### PR DESCRIPTION
**Pull request by alyssawilk**
Fixing a bug where upgrade response headers triggered the upgrade path

Risk Level: low (fixing a bug, runtime guarded)
Testing: fixed unit tests
Docs Changes: n/a
Release Notes: yes
Fixes #10566

### CODE REVIEW

**Comment by mattklein123** [comment]
@alyssawilk I might be misunderstanding, but I don't think this fixes https://github.com/envoyproxy/envoy/issues/10566. The issue in that ticket AFIUI is that the client codec is shifting into upgrade mode when it received the upgrade header in the response, even though it isn't in response to a valid upgrade request. I think we need to actually remember in the request path if it's a valid upgrade request (like we do for HEAD requests) and then not shift into upgrade mode in the response in that case?

**Comment by alyssawilk** [comment]
Ah, I think there's 2 bugs.  There's what the codec does pulling in data, and what it does when serializing it out.  I was indeed fixing only one, making sure it was written out correctly, where we need to fix both.  Better fix and integration test coming up!

**Comment by asraa** [comment]
Clearing response upgrade headers here would also fix this too, right? https://github.com/envoyproxy/envoy/pull/8797

**Comment by alyssawilk** [comment]
This does clear them over in conn_manager_utility.cc
Did you mean clearing in the codec?

File: test/integration/integration_test.cc [hunk 1]
```diff
@@ -1227,4 +1227,29 @@ TEST_P(IntegrationTest, TestManyBadRequests) {
...
+  ASSERT(fake_upstream_connection != nullptr);
+  ASSERT_TRUE(fake_upstream_connection->write("HTTP/1.1 200 OK\r\n"
+                                              "connection: upgrade\r\n"
+                                              "upgrade: h2\r\n"
+                                              "Transfer-encoding: chunked\r\n\r\n"
+                                              "b\r\nHello World\r\n0\r\n\r\n",
+                                              false));
+
+  response->waitForHeaders();
+  EXPECT_EQ(nullptr, response->headers().Upgrade());
```
Comment (mattklein123) [inline test/integration/integration_test.cc:1248]: maybe verify the connection header was stripped also?

File: source/common/http/http1/codec_impl.cc [hunk 2]
```diff
@@ -949,6 +952,13 @@ int ClientConnectionImpl::onHeadersComplete() {
   return cannotHaveBody() ? 1 : 0;
 }
 
+bool ClientConnectionImpl::upgradeAllowed() const {
+  if (pending_response_.has_value()) {
+    return pending_response_->encoder_.upgrade_request_;
+  }
+  return false;
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc:959]: Do we have a test that hits this branch?

File: docs/root/intro/version_history.rst [hunk 3]
```diff
@@ -21,6 +21,10 @@ Version history
 * datasource: added retry policy for remote async data source.
 * dns: the STRICT_DNS cluster now only resolves to 0 hosts if DNS resolution successfully returns 0 hosts.
 * dns: added support for :ref:`dns_failure_refresh_rate <envoy_api_field_config.common.dynamic_forward_proxy.v2alpha.DnsCacheConfig.dns_failure_refresh_rate>` for the :ref:`dns cache <envoy_api_msg_config.common.dynamic_forward_proxy.v2alpha.DnsCacheConfig>` to set the DNS refresh rate during failures.
+* http filters: http filter extensions use the "envoy.filters.http" name space. A mapping
+  of extension names is available in the :ref:`deprecated <deprecated>` documentation.
+* http: fixing a bug where the upgrade header was not cleared on responses to non-upgrade requests.
+  Can be reverted temporarily by setting runtime feature `envoy.reloadable_features.fix_upgrade_response` to false.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst:27]: nit: alpha order issue.

Comment (alyssawilk) [reply docs/root/intro/version_history.rst]: Fixing it forward: https://github.com/envoyproxy/envoy/pull/10695 :-P

**Comment by mattklein123** [comment]
Sorry can you merge master and move to 1.15.0?

/wait

**Comment by alyssawilk** [comment]
@lambdai for ?backport?

File: docs/root/intro/version_history.rst [hunk 4]
```diff
@@ -4,6 +4,9 @@ Version history
 1.15.0 (Pending)
 ================
 
+* http: fixing a bug where the upgrade header was not cleared on responses to non-upgrade requests.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst:7]: s/fixing/fixed (per your own request =P)?

**Comment by mattklein123** [comment]
LGTM. Can you merge master in the morning once CI is fixed so we can make sure this passes?

/wait

**Comment by mattklein123** [comment]
docs CI error looks legit :(

/wait

**Pull request merged**